lint/base_linter/node_linter.py

```diff
@@ -146,21 +147,67 @@ def find_local_executable(self, start_dir, npm_name):
-                            .format(
-                                npm_name,
-                                'dependency' if is_dep else 'devDependency',
-                                manifest_file
+                is_dep = bool(manifest.get('dependencies', {}).get(npm_name))
+                is_dev_dep = bool(manifest.get('devDependencies', {}).get(npm_name))
+                if is_dep or is_dev_dep:
+                    self.context['project_root'] = path
+
+                    # Perhaps this is a Yarn PnP project?
+                    yarn_path = shutil.which('yarn')
+                    yarn_lock_exists = os.path.exists(os.path.join(path, 'yarn.lock'))
+                    package_lock_exists = os.path.exists(os.path.join(path, 'package-lock.json'))
+                    pnp_js_exists = os.path.exists(os.path.join(path, '.pnp.js'))
+                    pnp_in_manifest = bool(manifest.get('installConfig', {}).get('pnp'))
+
+                    if yarn_path is None:
+                        if yarn_lock_exists and (pnp_in_manifest or pnp_js_exists):
+                            logger.warning(
+                                "This seems like a Yarn PnP project. However, finding "
+                                "a Yarn executable failed. Make sure to install Yarn first."
                             )
+                            self.notify_failure()
+                            raise linter.PermanentError()
+                    elif yarn_lock_exists and pnp_js_exists:
+                        return [yarn_path, 'run', '--silent', npm_name]
+
+                    next_path = os.path.dirname(path)
+
+                    for path in paths_upwards(next_path):
+                        executable = shutil.which(npm_name, path=os.path.join(path, 'node_modules', '.bin'))
+                        if executable:
+                            return executable
+
+                    logger.warning(
+                        "Skipping '{}' for now which is listed as a {} "
+                        "in {} but not installed.  Forgot to '{} install'?"
+                        .format(
+                            npm_name,
+                            'dependency' if is_dep else 'devDependency',
+                            manifest_file,
+                            'yarn' if yarn_path and yarn_lock_exists and not package_lock_exists else 'npm'
                         )
-                        self.notify_failure()
-                        raise linter.PermanentError()
+                    )
+                    self.notify_failure()
+                    raise linter.PermanentError()
 
         return None
+
+    def run(self, cmd, code):
+        result = super().run(cmd, code)
+
+        try:
+            if ('error Command "' + cmd[3] + '" not found') in str(result):
+                logger.warning(
+                    "Skipping '{}' for now which is listed as a dependency/devDependency "
```

Comment:
Maybe: "We did run yarn run --silent cmd but the command cmd cannot be found.  Forgot to run ...." Then we don't have to think about a probable edge case that project_root is not set. I.e. if the user set executable manually.